src/lsm/tree.zig

```diff
@@ -266,6 +272,88 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
+        ///    index block, filter block & data block for the tables that could contain the key
+        ///    synchronously from the Grid cache. If any of these blocks aren't present in the cache,
+        ///    it returns True since the existence of the key cannot be determined.
+        pub fn key_may_exist(tree: *Tree, key: Key, location: union(enum) { tree, manifest }) bool {
+            assert((tree.key_min == null and tree.key_max == null) or
+                (tree.key_min != null and tree.key_max != null));
+            switch (location) {
+                .tree => {
+                    // Tree key range hasn't been updated yet.
+                    if (tree.key_min == null and tree.key_max == null) {
+                        return false;
+                    }
+                    if (compare_keys(key, tree.key_min.?) == .lt or compare_keys(key, tree.key_max.?) == .gt) {
+                        return false;
+                    } else {
+                        return true;
+                    }
+                },
+                .manifest => {
+                    var iterator = tree.manifest.key_may_exist(key);
+                    if (iterator.next() == null) {
+                        return false;
+                    }
+                    const fingerprint = bloom_filter.Fingerprint.create(stdx.hash_inline(key));
+                    while (iterator.next()) |table| {
+                        if (tree.grid.read_from_cache(table.address, table.checksum)) |index_block| {
+                            const blocks = Table.index_blocks_for_key(index_block, key);
+                            if (tree.grid.read_from_cache(
+                                blocks.filter_block_address,
+                                blocks.filter_block_checksum,
+                            )) |filter_block| {
+                                const filter_bytes = Table.filter_block_filter_const(filter_block);
+                                if (!bloom_filter.may_contain(fingerprint, filter_bytes)) {
+                                    continue;
+                                }
+                            } else {
+                                // Filter block not in cache. We could use either data block or
+                                // filter block to determine existence of a key within a table, so
+                                // we attempt to search for data block in the cache.
+                            }
+                            if (tree.grid.read_from_cache(
+                                blocks.data_block_address,
+                                blocks.data_block_checksum,
+                            )) |data_block| {
+                                if (Table.data_block_search(data_block, key)) |value| {
+                                    if (unwrap_tombstone(value) == null) return false;
+                                    return true;
+                                }
```

Comment:
If the key is not found in the data block, do we not also expect a continue here in this else branch?

Reply:
Added an explicit `continue` statement, for clarity!

Reply:
I have made this explicit with an `else` clause now, so that the two branches (for filter & data block) are symmetrical.